## Date : 06/10/2025

In [2]:
pip install pandas 

Note: you may need to restart the kernel to use updated packages.


In [1]:
import pandas as pd

# series - Series is 1-D Array / Single column which is indexed
series = pd.Series([2,3,5,4,6], index = [1,2,3,4,5])
print(series)

1    2
2    3
3    5
4    4
5    6
dtype: int64


In [6]:
# dataframe - dataframe is 2-D Array / the data is stored form as row's and column's
# featues = column's
# instance/data points = row's

dataframe = {
    "name":["asd","fds","rew","vsds","sdf"],
    "score":[88,23,45,45,12],
    #"age":[23,54,12,45,63]
}
df = pd.DataFrame(dataframe)
print(df)

   name  score
0   asd     88
1   fds     23
2   rew     45
3  vsds     45
4   sdf     12


In [2]:
import pandas as pd
df = pd.read_csv("F://top_1000_popular_movies_tmdb.csv", engine = "python")

In [3]:
df.head() # head() return frist 5 row's of the dataset by default , in case if you want any other number of row's pass the number inside in head()

Unnamed: 0      id                                title release_date  \
0          0  385687                               Fast X   2023-05-17   
1          1  603692                 John Wick: Chapter 4   2023-03-22   
2          2  502356          The Super Mario Bros. Movie   2023-04-05   
3          3  569094  Spider-Man: Across the Spider-Verse   2023-05-31   
4          4  536437                             Hypnotic   2023-05-11   

                                              genres original_language  \
0                    ['Action', 'Crime', 'Thriller']           English   
1                    ['Action', 'Thriller', 'Crime']           English   
2  ['Animation', 'Family', 'Adventure', 'Fantasy'...           English   
3  ['Action', 'Adventure', 'Animation', 'Science ...           English   
4         ['Mystery', 'Thriller', 'Science Fiction']           English   

   vote_average  vote_count  popularity  \
0           7.4      1347.0    8363.473   
1           7.9      2896.0    4210.313   
2           7.8      4628.0    3394.458   
3           8.8      1160.0    2859.047   
4           6.5       154.0    2654.854   

                                            overview       budget  \
0  Over many missions and against impossible odds...  340000000.0   
1  With the price on his head ever increasing, Jo...   90000000.0   
2  While working underground to fix a water main,...  100000000.0   
3  After reuniting with Gwen Stacy, Brooklyn’s fu...  100000000.0   
4  A detective becomes entangled in a mystery inv...   70000000.0   

                                production_companies       revenue  runtime  \
0  ['Universal Pictures', 'Original Film', 'One R...  6.520000e+08    142.0   
1  ['Thunder Road', '87Eleven', 'Summit Entertain...  4.317692e+08    170.0   
2  ['Universal Pictures', 'Illumination', 'Ninten...  1.308767e+09     92.0   
3  ['Columbia Pictures', 'Sony Pictures Animation...  3.135222e+08    140.0   
4  ['Studio 8', 'Solstice Productions', 'Ingeniou...  0.000000e+00     94.0   

                                   tagline  
0              The end of the road begins.  
1                No way back, one way out.  
2                                      NaN  
3  It's how you wear the mask that matters  
4                  Control is an illusion.

In [4]:
df.tail() # tail()  returns the last 5 rows of the data set by default , for any other number pass it inside tail method tail()

Unnamed: 0      id                           title release_date  \
9996        9995   33142                           Vivah   2006-11-10   
9997        9996    9442       Dead Men Don't Wear Plaid   1982-05-21   
9998        9997   28774                       Communion   1989-11-10   
9999        9998   11121                            Tess   1979-10-06   
10000       9999  272692  The Possession of Michael King   2014-08-14   

                                                  genres original_language  \
9996                                ['Romance', 'Drama']             Hindi   
9997                               ['Comedy', 'Mystery']           English   
9998   ['Drama', 'Horror', 'Science Fiction', 'Thrill...           English   
9999                                ['Drama', 'Romance']           English   
10000                                         ['Horror']           English   

       vote_average  vote_count  popularity  \
9996            6.4        70.0      12.301   
9997            6.6       324.0      12.301   
9998            5.3       147.0      12.300   
9999            7.0       283.0      12.300   
10000           6.1       346.0      12.297   

                                                overview      budget  \
9996   Poonam, a traditionally brought-up young woman...         0.0   
9997   Juliet Forrest is convinced that the reported ...   9000000.0   
9998   A novelist's wife and son see him changed by a...   5000000.0   
9999   A strong-willed peasant girl is sent by her fa...  12000000.0   
10000  The film tells the story of documentary filmma...         0.0   

                                    production_companies     revenue  runtime  \
9996                             ['Rajshri Productions']   4000000.0    175.0   
9997        ['Universal Pictures', 'Aspen Film Society']  18196170.0     89.0   
9998   ['The Picture Property Company', 'Allied Visio...   1919653.0    103.0   
9999   ['Renn Productions', 'Timothy Burrill Producti...  20101247.0    186.0   
10000           ['Quickfire Films', 'Gold Circle Films']   2405143.0     83.0   

                                                 tagline  
9996                            Marriage isn't a game...  
9997                 Laugh...or I’ll blow your lips off!  
9998   The true story of one man's terrifying journey...  
9999   She was born into a world where they called it...  
10000     If you invite it in, it will never let you go.

In [5]:
df.info() # info() returns information of all the columns 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10001 entries, 0 to 10000
Data columns (total 15 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Unnamed: 0            10001 non-null  object 
 1   id                    10001 non-null  int64  
 2   title                 10001 non-null  object 
 3   release_date          9978 non-null   object 
 4   genres                10001 non-null  object 
 5   original_language     10000 non-null  object 
 6   vote_average          10000 non-null  float64
 7   vote_count            10000 non-null  float64
 8   popularity            10000 non-null  float64
 9   overview              9924 non-null   object 
 10  budget                9999 non-null   float64
 11  production_companies  9999 non-null   object 
 12  revenue               9999 non-null   float64
 13  runtime               9999 non-null   float64
 14  tagline               7383 non-null   object 
dtypes: float64(6), int6

In [6]:
df.describe() # describe() will return basic stastical summary of all numberical colums

id  vote_average    vote_count    popularity        budget  \
count  1.000100e+04  10000.000000  10000.000000  10000.000000  9.999000e+03   
mean   3.001465e+05      6.319090   1558.294800     33.542340  1.993800e+07   
std    3.379901e+05      1.460605   2887.861217    126.495621  3.870582e+07   
min    0.000000e+00      0.000000      0.000000     12.297000  0.000000e+00   
25%    1.150400e+04      5.900000    143.000000     14.599000  0.000000e+00   
50%    1.184060e+05      6.600000    519.000000     18.582500  1.250000e+06   
75%    5.379150e+05      7.100000   1584.000000     28.014500  2.300000e+07   
max    1.136631e+06     10.000000  33822.000000   8363.473000  4.600000e+08   

            revenue      runtime  
count  9.999000e+03  9999.000000  
mean   6.084337e+07   100.809581  
std    1.553073e+08    27.850165  
min    0.000000e+00     0.000000  
25%    0.000000e+00    90.000000  
50%    2.133452e+06   100.000000  
75%    5.106230e+07   115.000000  
max    2.923706e+09   366.000000

In [7]:
df.describe(include = "all") # describe(include = "all") return the stastical summary of all colums

Unnamed: 0            id               title release_date     genres  \
count       10001  1.000100e+04               10001         9978      10001   
unique      10001           NaN                9654         5996       2280   
top          9999           NaN  The Little Mermaid   2023-06-11  ['Drama']   
freq            1           NaN                   5           32        582   
mean          NaN  3.001465e+05                 NaN          NaN        NaN   
std           NaN  3.379901e+05                 NaN          NaN        NaN   
min           NaN  0.000000e+00                 NaN          NaN        NaN   
25%           NaN  1.150400e+04                 NaN          NaN        NaN   
50%           NaN  1.184060e+05                 NaN          NaN        NaN   
75%           NaN  5.379150e+05                 NaN          NaN        NaN   
max           NaN  1.136631e+06                 NaN          NaN        NaN   

       original_language  vote_average    vote_count    popularity  \
count              10000  10000.000000  10000.000000  10000.000000   
unique                52           NaN           NaN           NaN   
top              English           NaN           NaN           NaN   
freq                7304           NaN           NaN           NaN   
mean                 NaN      6.319090   1558.294800     33.542340   
std                  NaN      1.460605   2887.861217    126.495621   
min                  NaN      0.000000      0.000000     12.297000   
25%                  NaN      5.900000    143.000000     14.599000   
50%                  NaN      6.600000    519.000000     18.582500   
75%                  NaN      7.100000   1584.000000     28.014500   
max                  NaN     10.000000  33822.000000   8363.473000   

                    overview        budget production_companies       revenue  \
count                   9924  9.999000e+03                 9999  9.999000e+03   
unique                  9920           NaN                 8035           NaN   
top     Nikkatsu Roman Porno           NaN                   []           NaN   
freq                       3           NaN                  518           NaN   
mean                     NaN  1.993800e+07                  NaN  6.084337e+07   
std                      NaN  3.870582e+07                  NaN  1.553073e+08   
min                      NaN  0.000000e+00                  NaN  0.000000e+00   
25%                      NaN  0.000000e+00                  NaN  0.000000e+00   
50%                      NaN  1.250000e+06                  NaN  2.133452e+06   
75%                      NaN  2.300000e+07                  NaN  5.106230e+07   
max                      NaN  4.600000e+08                  NaN  2.923706e+09   

            runtime                 tagline  
count   9999.000000                    7383  
unique          NaN                    7328  
top             NaN  Based on a true story.  
freq            NaN                       6  
mean     100.809581                     NaN  
std       27.850165                     NaN  
min        0.000000                     NaN  
25%       90.000000                     NaN  
50%      100.000000                     NaN  
75%      115.000000                     NaN  
max      366.000000                     NaN

**1.Title -** 
- **Count** The title feature has **10,001** values, indicating that there are no missing values in this column.
- **Unique Values** There are **9,645** unique titles, meaning that most movies have distinct titles.
- **Most Frequent Value** The title **"The Little Mermaid"** is the most frequent, appearing 5 times in the dataset.

**2. Release date -**
- **Count** The release_date feature has **9,978** non-null values, indicating that there are some missing values in the dataset.
- **Unique Values** There are **5,996** unique release dates, showing that many movies share the same release date.
- **Most Frequent Value** The release date **"2023-06-11"** is the most common, appearing 32 times in the dataset.

**3. Genres -**
- **Count** The genres feature has **10,001** values, meaning there are no missing values in this feature.
- **Unique Values** There are **2,280** unique genre entries, indicating a diverse range of genre combinations across the movies.
- **Most Frequent Value** The genre **"Drama"** is the most common, appearing 582 times in the dataset.

**4. Original language -**
- **Count** The original_language feature has **10,000** values, meaning there is **1 missing** value in the dataset.
- **Unique Values** There are **52** unique languages represented in the dataset, showing a wide variety of original languages for the movies.
- **Most Frequent Value** The language **"English"** is the most common, appearing **7,340** times in the dataset.

**5. Vote average -**
- **Count** The vote_average feature has **10,000** values, indicating no missing values 
- **Mean** The average vote rating is approximately **6.32**, which suggests that movies generally receive moderately positive ratings.
- **Standard Deviation** The standard deviation is **1.46**, showing a relatively small variation in vote averages across movies.
- **Minimum** The lowest vote average is **0.00**, which may represent unrated or erroneous entries.
- **25th Percentile (Q1)** 25% of the movies have a vote average less than or equal to **5.90**, indicating a quarter of the movies received below-average ratings.
- **Median (50th Percentile)** Half of the movies have a vote average less than or equal to **6.60**, meaning the typical movie rating is slightly above 6.
- **75th Percentile (Q3)** 75% of the movies have vote averages below **7.10**, while the top 25% are rated higher.
- **Maximum** The highest vote average is **10.00**, indicating perfect ratings for at least one movie in the dataset.

**6.Vote count** -
- **Count** The vote_count feature has **10,000** values, which means no missing values.
- **Mean** The average number of votes is approximately **1,558.29**, suggesting that, on average, movies receive over 1,500 user ratings.
- **Standard Deviation** The standard deviation is **2,887.86**, indicating a large spread in vote counts — some movies are far more popular or rated than others.
- **Minimum** The lowest vote count is **0**, indicating that some movies have no recorded votes, possibly due to lack of audience engagement or missing data.
- **25th Percentile (Q1)** 25% of the movies have **143** or fewer votes, reflecting that a quarter of the dataset includes lesser-known or less-rated films.
- **Median (50th Percentile)** Half of the movies have fewer than or equal to **519** votes, meaning typical movies receive relatively moderate engagement.
- **75th Percentile (Q3)** 75% of the movies have fewer than or equal to **1,584** votes, while the top 25% of movies received more than that.
- **Maximum** The highest vote count is **33,822**, indicating the presence of very popular or widely rated movies in the dataset.

**7. Popularity -**
- **Count** The popularity feature has **10,000** values, which indicates no missing values.
- **Mean** The average popularity score is approximately **33.54**, suggesting that most movies have relatively low to moderate popularity, with a few high-scoring outliers.
- **Standard Deviation** The standard deviation is **126.50**, showing a very wide variation in popularity scores — likely due to a few movies being significantly more popular than the rest.
- **Minimum** The lowest popularity score is **12.30**, which likely reflects lesser-known or niche films.
- **25th Percentile (Q1)** 25% of the movies have a popularity score below **14.60**, indicating that a quarter of the dataset consists of low-popularity movies.
- **Median (50th Percentile)** Half of the movies have a popularity score below **18.58**, suggesting that the typical movie has moderate popularity.
- **75th Percentile (Q3)** 75% of the movies have a popularity score below **28.01**, and the top 25% exceed this value.
- **Maximum** The highest popularity score is 8,363.47, indicating a major outlier — a movie that was extremely popular compared to the rest.

**8. Overview -**
- **Count** The overview feature has **9,924** non-null values, indicating that there are some missing values in the dataset.
- **Unique Values** There are **9,920** unique overviews, showing that most movies have a distinct description, with only a few repetitions.
- **Most Frequent Value** The phrase **"Nikkatsu Roman Porno"** appears **3 times**, making it the most repeated overview in the dataset.


**9. Production companies -**
- **Count** The original_language feature has **9,999** non-null values, which means there is 1 missing value in the dataset.
- **Unique Values** There are **8,035** unique language entries, indicating a wide variety of original languages used across the movies.
- **Most Frequent Value** The missing value **null** is recorded as the most frequent, appearing **518** times, suggesting possible data entry issues or missing language information.

**10. Revenue -**
- **Count** The revenue feature has **9,999** values, indicating that there is 1 missing value in the dataset.
- **Mean** The average revenue is approximately **RS.60,843,370.00**, 
- **Standard Deviation** The standard deviation is **RS.155,307,300.00**, which reflects a very high spread in revenue values — suggesting the presence of significant outliers.
- **Minimum** The lowest revenue is **RS.0.00**, indicating that some movies either did not earn revenue or the value was not reported.
- **25th Percentile (Q1)** 25% of the movies have a revenue of **RS.0.00**, further confirming that a large portion of entries are missing or have no revenue reported.
- **Median (50th Percentile)** Half of the movies earned less than or equal to **RS.2,133,452.00**, showing that most movies earned relatively low revenue.
- **75th Percentile (Q3)** 75% of the movies have revenue below **RS.51,062,300.00**, while the top 25% earned more than this amount.
- **Maximum** The highest revenue recorded is **RS.2,923,706,000.00**, indicating the presence of massively successful box-office hits in the dataset.

**11. Budget -**
- **Count** The budget feature has **9,999** values, indicating that 1 value is missing from the dataset.
- **Mean** The average budget across all movies is **RS.19,938,000.00.**
- **Standard Deviation** The standard deviation is **RS.38,705,820.00,** showing a large spread in budget values, suggesting high variability.
- **Minimum** The smallest budget recorded is **RS,0.00,** which may represent missing or unreported data.
- **25th Percentile (Q1)** 25% of the movies have a budget of **RS.0.00** or less, indicating that at least one-quarter of the entries either had no budget or it was not reported.
- **Median (50th Percentile)** Half of the movies have a budget less than or equal to **RS.1,250,000.00,** showing that the typical movie has a relatively low budget.
- **75th Percentile (Q3)** 75% of the movies have budgets below **RS.23,000,000.00.** The top 25% of movies had budgets higher than this.
- **Maximum** The highest budget recorded is RS.460,000,000.00,** indicating the presence of extreme outliers with very high budgets.

**12.Runtime -**
- **Count** The runtime feature has **9,999** values, which means 1 value is missing in the dataset.
- **Mean** The average runtime of movies is approximately **100.81** minutes.
- **Standard Deviation** The standard deviation is **27.85** minutes, indicating a moderate spread in movie lengths.
- **Minimum** The shortest runtime is **0.00** minutes, which likely represents a missing or incorrect entry.
- **25th Percentile (Q1)** 25% of the movies have a runtime less than or equal to **90 minutes**, suggesting a significant portion of movies are relatively short.
- **Median (50th Percentile)** Half of the movies have a runtime less than or equal to **100 minutes**, indicating that most movies are around this length.
- **75th Percentile (Q3)** 75% of the movies have runtimes below **115 minutes**, and the top 25% are longer than that.
- **Maximum** The longest runtime is **366 minutes**, indicating the presence of outliers or possibly multi-part or special edition films.

**13. Tagline -**
- **Count** The tagline feature has **7,383** non-null values, indicating that there are some missing values in this column.
- **Unique Values** There are **7,328** unique taglines, showing that most movies have distinct taglines, with very few repetitions.
- Most Frequent Value: The tagline **"Based on a true story."** is the most common, appearing **6 times** in the dataset.

In [2]:
import pandas as pd
df = pd.read_csv("F://top_1000_popular_movies_tmdb.csv", engine = "python")

In [4]:
df.shape - # Shape return number of row's and column's in the data set.

SyntaxError: invalid syntax (4032481966.py, line 1)

In [10]:
df.columns # columns return all calumns in the data set with name's.

Index(['Unnamed: 0', 'id', 'title', 'release_date', 'genres',
       'original_language', 'vote_average', 'vote_count', 'popularity',
       'overview', 'budget', 'production_companies', 'revenue', 'runtime',
       'tagline'],
      dtype='object')

In [11]:
df.index # index return range of the row's in the data set.

RangeIndex(start=0, stop=10001, step=1)

In [3]:
df[df['release_date'].isnull()] 

Unnamed: 0       id                                            title  \
1073       1072   626412                                 Alienoid: Part 2   
1299       1298  1034541                                      Terrifier 3   
1538       1537   421892                                          Shrek 5   
1594       1593   969681                                     Spider-Man 4   
1643       1642  1108427                                            Moana   
2217       2216  1126217                                   Fast X: Part 3   
2705       2704   631969                         Kingsman: The Blue Blood   
2738       2737   730629                             John Wick: Chapter 5   
3360       3359   504075                                   Wolf Warrior 3   
3443       3442   553301                                        Your Name   
4171       4170   628861                                    Baby Driver 2   
4951       4950   956421                              Save Me, Spider-Man   
5616       5615  1084242                                       Zootopia 2   
5715       5714   939345               Transformers: Rise of the Beasts 2   
5813       5812   939347               Transformers: Rise of the Beasts 3   
6872       6871  1135085        Untitled Miles Morales Spider-Man Project   
7528       7527   626741                          G.I. Joe: Ever Vigilant   
7908       7907   695727                                    One Punch Man   
7941       7940   347196                                     Wolf Creek 3   
8005       8004  1084244                                      Toy Story 5   
8364       8363  1074435  Red Bull Dragon Ball FighterZ World Final Paris   
8456       8455   574475                  Final Destination 6: Bloodlines   
9762       9761   391331                          El día de los albañiles   

     release_date                                        genres  \
1073          NaN                                            []   
1299          NaN             ['Horror', 'Fantasy', 'Thriller']   
1538          NaN  ['Animation', 'Family', 'Comedy', 'Fantasy']   
1594          NaN                                    ['Action']   
1643          NaN  ['Adventure', 'Comedy', 'Family', 'Fantasy']   
2217          NaN                                            []   
2705          NaN             ['Action', 'Comedy', 'Adventure']   
2738          NaN  ['Action', 'Adventure', 'Crime', 'Thriller']   
3360          NaN                             ['Action', 'War']   
3443          NaN               ['Drama', 'Fantasy', 'Romance']   
4171          NaN                           ['Action', 'Crime']   
4951          NaN                                            []   
5616          NaN             ['Animation', 'Comedy', 'Family']   
5715          NaN    ['Action', 'Adventure', 'Science Fiction']   
5813          NaN    ['Action', 'Adventure', 'Science Fiction']   
6872          NaN            ['Action', 'Fantasy', 'Adventure']   
7528          NaN    ['Action', 'Adventure', 'Science Fiction']   
7908          NaN            ['Action', 'Adventure', 'Fantasy']   
7941          NaN               ['Drama', 'Horror', 'Thriller']   
8005          NaN          ['Animation', 'Adventure', 'Family']   
8364          NaN                               ['Documentary']   
8456          NaN                         ['Horror', 'Mystery']   
9762          NaN                                    ['Comedy']   

     original_language  vote_average  vote_count  popularity  \
1073            Korean           0.0         0.0      47.439   
1299           English           0.0         0.0      41.450   
1538           English           0.0         0.0      37.458   
1594           English           0.0         0.0      36.681   
1643           English           0.0         0.0      35.976   
2217           English           0.0         0.0      30.089   
2705           English           0.0         0.0      26.741   
2738           English           0.

1073 -Alienoid: Part 2 - 10 January 2024 
1299 -Terrifier 3 - 19 September 2024
1538 - Shrek 5 -30 June 2027
1594 - July 31, 2026. 
1643 - 23 November 2016
2217 - not con
2705 - not con
2738 - not con
3360 - not con
3443 - 26 August 2016
4171 - not con
4951 - July 31, 2026
5616 - 26 November 2025
5715 - not con
5813 - not con
6872 - June 18, 2027
7528 - not con
7908 - not con
7941 - not con
8005 - 19 June 2026
8364 - not con
8456 - 16 May 2025
9762 - August 9, 1984.

In [71]:
"""
isnull() is used to check for missing values and it return the output in boolean which is hard to interpret hence it's used with sum()
to get the total count of missing values in each feature.
"""
df.isnull().sum() 

id                       0
title                    0
release_date            19
genres                   0
language                 0
vote_average             0
vote_count               0
popularity               0
budget                   0
production_companies     0
revenue                  0
runtime                  0
year                    19
dtype: int64

In [5]:
"""
isna() is used to check for missing values and it return the output in boolean which is hard to interpret hence it's used with sum()
to get the total count of missing values in each feature.
"""
df.isna().sum()

Unnamed: 0                 0
id                         0
title                      0
release_date              23
genres                     0
original_language          1
vote_average               1
vote_count                 1
popularity                 1
overview                  77
budget                     2
production_companies       2
revenue                    2
runtime                    2
tagline                 2618
dtype: int64

In [6]:
"""
loc() is used to locate the row which matches the criteria
"""

df.loc[df['release_date'].isna() & (df['title'] == 'Alienoid: Part 2'),'release_date'] = '2024/01/10'
df.loc[df['release_date'].isna() & (df['title'] == 'Terrifier 3'),'release_date'] = '2024/09/19'
df.loc[df['release_date'].isna() & (df['title'] == 'Shrek 5'),'release_date'] = '2027/06/30'
df.loc[df['release_date'].isna() & (df['title'] == 'Spider-Man 4'),'release_date'] = '2026/06/31'
df.loc[df['release_date'].isna() & (df['title'] == 'Moana'),'release_date'] = '2016/12/02'
df.loc[df['release_date'].isna() & (df['title'] == 'Fast X: Part 3'),'release_date'] = '2027/04/01'
df.loc[df['release_date'].isna() & (df['title'] == 'Wolf Warrior 3'),'release_date'] = '2025/05/01'
df.loc[df['release_date'].isna() & (df['title'] == 'Save Me, Spider-Man'),'release_date'] = '2027/06/18'
df.loc[df['release_date'].isna() & (df['title'] == 'Zootopia 2'),'release_date'] = '2025/11/26'
df.loc[df['release_date'].isna() & (df['title'] == 'Transformers: Rise of the Beasts 2'),'release_date'] = '2024/06/24'
df.loc[df['release_date'].isna() & (df['title'] == 'Transformers: Rise of the Beasts 3'),'release_date'] = '2023/06/09'
df.loc[df['release_date'].isna() & (df['title'] == 'Untitled Miles Morales Spider-Man Project'),'release_date'] = '2027/06/04'
df.loc[df['release_date'].isna() & (df['title'] == 'G.I. Joe: Ever Vigilant'),'release_date'] = '2025/07/18'
df.loc[df['release_date'].isna() & (df['title'] == 'One Punch Man'),'release_date'] = '2025/10/05'
df.loc[df['release_date'].isna() & (df['title'] == 'Toy Story 5'),'release_date'] = '2025/06/18'
df.loc[df['release_date'].isna() & (df['title'] == 'Final Destination 6: Bloodlines'),'release_date'] = '2025/05/16'
df.loc[df['release_date'].isna() & (df['title'] == 'El día de los albañiles'),'release_date'] = '1984/04/04'
df.loc[df['release_date'].isna() & (df['title'] == 'Red Bull Dragon Ball FighterZ World Final Paris'),'release_date'] = '2020/02/08'
df.loc[df['release_date'].isna() & (df['title'] == 'Your Name'),'release_date'] = '2016/08/26'

In [ ]:
release_date   movie name
10-1-2024    Alienoid: Part 2
4-4-2025     Terrifier 3
30-6-2027    Shrek 5	
31-7-2026   Spider-Man 4	
2-12-2016     Moana	
2027/04/01     Fast X: Part 3	
NA           Kingsman: The Blue Blood	  
[No official   John Wick: Chapter 5	
release date]
1-5-2025     Wolf Warrior 3	
26-8-2016    Your Name
2025/12/31      Baby Driver 2
18-6-2027    Save Me, Spider-Man	
26-11-2025    Zootopia 2	
24-6-2024   Transformers: Rise of the Beasts 2	
9-6-2023    Transformers: Rise of the Beasts 3	
4-6-2027    Untitled Miles Morales Spider-Man Project	
18-7-2025   G.I. Joe: Ever Vigilant	
5-10-2025     One Punch Man
NA       Wolf Creek 3	
2026/06/19   Toy Story 5	
2020/02/08  Red Bull Dragon Ball FighterZ World Final Paris	
16-5-2025    Final Destination 6: Bloodlines	
4-4-1984    El día de los albañiles

In [7]:
"""
dropna() is used to drop missing values in the columns we mention in subset, inplace = True ensures the changes are made to the actual dataframe.
"""
df.dropna(subset = 'release_date', inplace = True)

In [8]:
df['overview'] = df['overview'].fillna('unknown')

In [9]:
df['tagline'] = df['tagline'].fillna('No Tagline')

In [10]:
df.dropna(how = 'any', inplace = True)

In [11]:
df.shape

(9995, 15)

In [12]:
df.columns

Index(['Unnamed: 0', 'id', 'title', 'release_date', 'genres',
       'original_language', 'vote_average', 'vote_count', 'popularity',
       'overview', 'budget', 'production_companies', 'revenue', 'runtime',
       'tagline'],
      dtype='object')

In [13]:
df.drop(columns = ["Unnamed: 0","overview","tagline"], axis = 1 , inplace = True) 

In [14]:
df.head()

id                                title release_date  \
0  385687                               Fast X   2023-05-17   
1  603692                 John Wick: Chapter 4   2023-03-22   
2  502356          The Super Mario Bros. Movie   2023-04-05   
3  569094  Spider-Man: Across the Spider-Verse   2023-05-31   
4  536437                             Hypnotic   2023-05-11   

                                              genres original_language  \
0                    ['Action', 'Crime', 'Thriller']           English   
1                    ['Action', 'Thriller', 'Crime']           English   
2  ['Animation', 'Family', 'Adventure', 'Fantasy'...           English   
3  ['Action', 'Adventure', 'Animation', 'Science ...           English   
4         ['Mystery', 'Thriller', 'Science Fiction']           English   

   vote_average  vote_count  popularity       budget  \
0           7.4      1347.0    8363.473  340000000.0   
1           7.9      2896.0    4210.313   90000000.0   
2           7.8      4628.0    3394.458  100000000.0   
3           8.8      1160.0    2859.047  100000000.0   
4           6.5       154.0    2654.854   70000000.0   

                                production_companies       revenue  runtime  
0  ['Universal Pictures', 'Original Film', 'One R...  6.520000e+08    142.0  
1  ['Thunder Road', '87Eleven', 'Summit Entertain...  4.317692e+08    170.0  
2  ['Universal Pictures', 'Illumination', 'Ninten...  1.308767e+09     92.0  
3  ['Columbia Pictures', 'Sony Pictures Animation...  3.135222e+08    140.0  
4  ['Studio 8', 'Solstice Productions', 'Ingeniou...  0.000000e+00     94.0

In [15]:
df.reset_index(drop = True , inplace = True)

In [16]:
df.dtypes

id                        int64
title                    object
release_date             object
genres                   object
original_language        object
vote_average            float64
vote_count              float64
popularity              float64
budget                  float64
production_companies     object
revenue                 float64
runtime                 float64
dtype: object

In [17]:

df['release_date'] = pd.to_datetime(df["release_date"], errors = "coerce")

In [18]:
df['year'] = df['release_date'].dt.year

In [19]:
df['year'].head(10)

0    2023.0
1    2023.0
2    2023.0
3    2023.0
4    2023.0
5    2023.0
6    2023.0
7    2023.0
8    2022.0
9    2023.0
Name: year, dtype: float64

In [20]:
df.head(1072)

id                                title release_date  \
0     385687                               Fast X   2023-05-17   
1     603692                 John Wick: Chapter 4   2023-03-22   
2     502356          The Super Mario Bros. Movie   2023-04-05   
3     569094  Spider-Man: Across the Spider-Verse   2023-05-31   
4     536437                             Hypnotic   2023-05-11   
...      ...                                  ...          ...   
1067   18239          The Twilight Saga: New Moon   2009-11-18   
1068  262500                            Insurgent   2015-03-18   
1069   10009                         Brother Bear   2003-10-23   
1070    4256                        Scary Movie 3   2003-10-24   
1071  626412                     Alienoid: Part 2          NaT   

                                                 genres original_language  \
0                       ['Action', 'Crime', 'Thriller']           English   
1                       ['Action', 'Thriller', 'Crime']           English   
2     ['Animation', 'Family', 'Adventure', 'Fantasy'...           English   
3     ['Action', 'Adventure', 'Animation', 'Science ...           English   
4            ['Mystery', 'Thriller', 'Science Fiction']           English   
...                                                 ...               ...   
1067       ['Adventure', 'Fantasy', 'Drama', 'Romance']           English   
1068  ['Action', 'Adventure', 'Science Fiction', 'Th...           English   
1069               ['Adventure', 'Animation', 'Family']           English   
1070                                         ['Comedy']           English   
1071                                                 []            Korean   

      vote_average  vote_count  popularity       budget  \
0              7.4      1347.0    8363.473  340000000.0   
1              7.9      2896.0    4210.313   90000000.0   
2              7.8      4628.0    3394.458  100000000.0   
3              8.8      1160.0    2859.047  100000000.0   
4              6.5       154.0    2654.854   70000000.0   
...            ...         ...         ...          ...   
1067           6.0      8471.0      47.601   50000000.0   
1068           6.4      9369.0      47.575  110000000.0   
1069           7.3      4901.0      47.495  128000000.0   
1070           6.0      3646.0      47.486   48000000.0   
1071           0.0         0.0      47.439          0.0   

                                   production_companies       revenue  \
0     ['Universal Pictures', 'Original Film', 'One R...  6.520000e+08   
1     ['Thunder Road', '87Eleven', 'Summit Entertain...  4.317692e+08   
2     ['Universal Pictures', 'Illumination', 'Ninten...  1.308767e+09   
3     ['Columbia Pictures', 'Sony Pictures Animation...  3.135222e+08   
4     ['Studio 8', 'Solstice Productions', 'Ingeniou...  0.000000e+00   
...                                                 ...           ...   
1067  ['Summit Entertainment', 'Temple Hill Entertai...  7.098275e+08   
1068  ['Summit Entertainment', 'Red Wagon Entertainm...  2.952382e+08   
1069                  ['Walt Disney Feature Animation']  2.503978e+08   
1070          ['Brad Grey Pictures', 'Dimension Films']  2.206732e+08   
1071                                     ['Caper Film']  0.000000e+00   

      runtime    year  
0       142.0  2023.0  
1       170.0  2023.0  
2        92.0  2023.0  
3       140.0  2023.0  
4        94.0  2023.0  
...       ...     ...  
1067    131.0  2009.0  
1068    119.0  2015.0  
1069     85.0  2003.0  
1070     84.0  2003.0  
1071      0.0     NaN  

[1072 rows x 13 columns]

## Data Inspection Methods in Pandas :

In [21]:
"""
1.dtypes -
Shows the data type of each column in the DataFrame (e.g., int64, object, float64).
"""
df.dtypes

id                               int64
title                           object
release_date            datetime64[ns]
genres                          object
original_language               object
vote_average                   float64
vote_count                     float64
popularity                     float64
budget                         float64
production_companies            object
revenue                        float64
runtime                        float64
year                           float64
dtype: object

In [22]:
"""
2.columns -
Returns an Index object containing the names of all columns in the DataFrame.

"""
df.columns

Index(['id', 'title', 'release_date', 'genres', 'original_language',
       'vote_average', 'vote_count', 'popularity', 'budget',
       'production_companies', 'revenue', 'runtime', 'year'],
      dtype='object')

In [23]:
"""
3.index -
Returns the index (row labels) of the DataFrame.

"""
df.index

RangeIndex(start=0, stop=9995, step=1)

In [24]:
"""
4.isnull() - Returns a DataFrame of the same shape as df, with True for missing values (NaN) and False otherwise.

"""
df.isnull()

id  title  release_date  genres  original_language  vote_average  \
0     False  False         False   False              False         False   
1     False  False         False   False              False         False   
2     False  False         False   False              False         False   
3     False  False         False   False              False         False   
4     False  False         False   False              False         False   
...     ...    ...           ...     ...                ...           ...   
9990  False  False         False   False              False         False   
9991  False  False         False   False              False         False   
9992  False  False         False   False              False         False   
9993  False  False         False   False              False         False   
9994  False  False         False   False              False         False   

      vote_count  popularity  budget  production_companies  revenue  runtime  \
0          False       False   False                 False    False    False   
1          False       False   False                 False    False    False   
2          False       False   False                 False    False    False   
3          False       False   False                 False    False    False   
4          False       False   False                 False    False    False   
...          ...         ...     ...                   ...      ...      ...   
9990       False       False   False                 False    False    False   
9991       False       False   False                 False    False    False   
9992       False       False   False                 False    False    False   
9993       False       False   False                 False    False    False   
9994       False       False   False                 False    False    False   

       year  
0     False  
1     False  
2     False  
3     False  
4     False  
...     ...  
9990  False  
9991  False  
9992  False  
9993  False  
9994  False  

[9995 rows x 13 columns]

In [25]:
"""
5.isnull().sum() - Returns total number of missing values in each column

"""
df.isnull().sum()


id                       0
title                    0
release_date            19
genres                   0
original_language        0
vote_average             0
vote_count               0
popularity               0
budget                   0
production_companies     0
revenue                  0
runtime                  0
year                    19
dtype: int64

In [26]:
"""
6.duplicated() - Returns a boolean Series indicating whether each row is a duplicate of a previous row.

"""
df.duplicated()

0       False
1       False
2       False
3       False
4       False
        ...  
9990    False
9991    False
9992    False
9993    False
9994    False
Length: 9995, dtype: bool

In [27]:
"""
7.duplicated().sum() - Returns the total number of duplicate rows.

"""
df.duplicated().sum()


np.int64(0)

In [28]:
"""
8.nunique() - Returns the number of unique values per column.
"""
df.nunique()

id                      9995
title                   9648
release_date            5995
genres                  2279
original_language         52
vote_average              73
vote_count              3471
popularity              7965
budget                   699
production_companies    8031
revenue                 5372
runtime                  216
year                     108
dtype: int64

In [72]:
"""
9.value_counts() - Returns value counts for a Series (not DataFrame) — useful on a single column.

"""
df.value_counts('title')


title
Porno                             4
Masquerade                        3
Swingers                          3
All Quiet on the Western Front    2
Heidi                             2
                                 ..
О. А                              1
丈夫去上班的日子里                         1
日テレプラス presents 歌っ娘ライブ2 完全版       1
極道記者2 馬券転生篇                       1
Zatch Bell! 101st Devil           1
Name: count, Length: 4333, dtype: int64

In [30]:
"""
10.columns.tolist() - 
Converts the Index of column names to a Python list.
"""
df.columns.tolist()

['id',
 'title',
 'release_date',
 'genres',
 'original_language',
 'vote_average',
 'vote_count',
 'popularity',
 'budget',
 'production_companies',
 'revenue',
 'runtime',
 'year']

In [31]:
"""
11.index.tolist() -
Converts the Index of row labels to a list.
"""
#df.index.tolist()

'\nindex.tolist() -\nConverts the Index of row labels to a list.\n'

In [32]:
"""
12.empty -
Returns True if the DataFrame is empty (no rows or columns).
"""
df.empty

False

In [33]:
"""
13.sample(n) -
Randomly selects n rows from the DataFrame. Good for spot-checking data.
"""
df.sample(2)

id                 title release_date  \
5638  236399  Flowers in the Attic   2014-01-18   
2671   11811            Leprechaun   1993-01-08   

                                            genres original_language  \
5638  ['Drama', 'Mystery', 'Thriller', 'TV Movie']           English   
2671               ['Horror', 'Comedy', 'Fantasy']           English   

      vote_average  vote_count  popularity     budget  \
5638           6.7       324.0      17.352        0.0   
2671           5.6       662.0      26.898  1000000.0   

                                   production_companies    revenue  runtime  \
5638  ['Cue the Dog Productions', 'Front Street Pict...        0.0     90.0   
2671                               ['Trimark Pictures']  8556940.0     91.0   

        year  
5638  2014.0  
2671  1993.0

In [34]:
"""
14.Fetch one column
df['column name']
Returns the column specified in the form of series
"""
df["title"]

0                                    Fast X
1                      John Wick: Chapter 4
2               The Super Mario Bros. Movie
3       Spider-Man: Across the Spider-Verse
4                                  Hypnotic
                       ...                 
9990                                  Vivah
9991              Dead Men Don't Wear Plaid
9992                              Communion
9993                                   Tess
9994         The Possession of Michael King
Name: title, Length: 9995, dtype: object

In [35]:
"""
15.Fetch Multiple Columns
df['column1','column2']
Returns multiple specified columns in the form of dataframe.
"""
df[['title','budget']] 

title       budget
0                                  Fast X  340000000.0
1                    John Wick: Chapter 4   90000000.0
2             The Super Mario Bros. Movie  100000000.0
3     Spider-Man: Across the Spider-Verse  100000000.0
4                                Hypnotic   70000000.0
...                                   ...          ...
9990                                Vivah          0.0
9991            Dead Men Don't Wear Plaid    9000000.0
9992                            Communion    5000000.0
9993                                 Tess   12000000.0
9994       The Possession of Michael King          0.0

[9995 rows x 2 columns]

In [73]:
"""
16.iloc
df.iloc['row_label'] 
df.iloc['row_index, column_index']
Returns the row as per the specified row index.
"""
df.iloc[[457,3]]


id                           title release_date  \
1084  580116                        The Rape   1982-05-15   
15    879444  Accident Man: Hitman's Holiday   2022-10-14   

                             genres  language  vote_average  vote_count  \
1084                      ['Drama']  JAPANESE           3.5         5.0   
15    ['Action', 'Comedy', 'Crime']   ENGLISH           6.6        79.0   

      popularity  budget production_companies  revenue  runtime    year  
1084      46.993     0.0         ['Gentosha']      0.0    100.0  1982.0  
15      1117.559     0.0  ['Six Demon Films']      0.0     96.0  2022.0

In [38]:
'''
17..at
df.at[row_label, 'column_name']
Returns a single specified cell based on the specified row label, and column name.
'''
df.at[5,'original_language']

'English'

In [39]:
'''
18.iat
df.iat[row_index, column_index]
Returns a single specified cell based on specified row and column index
'''
df.iat[5,8]

np.float64(200000000.0)

In [40]:
'''
19.get()
df.get(key, default=None)
key: The column name you want.

default: Value to return if the column is not found. (Default is None)
It's similar to df['column_name'], but it won’t give an error if the column doesn’t exist — it just returns None (or whatever default you provide).
'''
df.get(key = 'original_language',default = None)

0       English
1       English
2       English
3       English
4       English
         ...   
9990      Hindi
9991    English
9992    English
9993    English
9994    English
Name: original_language, Length: 9995, dtype: object

In [41]:
'''
20.filter()
df.filter(items=None, like=None, regex=None, axis=None)
items:	List of names to keep
like:	Substring to match in labels (e.g. all columns with 'age' in the name)
regex:	Regular expression pattern to match labels
axis:	Axis to filter on: 0 = rows, 1 = columns (default is 1)
It is used to select specific rows or columns from a DataFrame by names or patterns.

- items =  Select only 'name' and 'salary' columns
filtered = df.filter(items=['name', 'salary'])
print(filtered)

- like = Select all columns that have 'age' in their name
filtered = df.filter(like='age')
print(filtered)

- regex = Select all columns that start with the letter 's'
filtered = df.filter(regex='^s')
print(filtered)

-items, axis = Select rows with index labels 0 and 2
filtered = df.filter(items=[0, 2], axis=0)
print(filtered)

'''
df.filter(items = ['title','genres'])


title  \
0                                  Fast X   
1                    John Wick: Chapter 4   
2             The Super Mario Bros. Movie   
3     Spider-Man: Across the Spider-Verse   
4                                Hypnotic   
...                                   ...   
9990                                Vivah   
9991            Dead Men Don't Wear Plaid   
9992                            Communion   
9993                                 Tess   
9994       The Possession of Michael King   

                                                 genres  
0                       ['Action', 'Crime', 'Thriller']  
1                       ['Action', 'Thriller', 'Crime']  
2     ['Animation', 'Family', 'Adventure', 'Fantasy'...  
3     ['Action', 'Adventure', 'Animation', 'Science ...  
4            ['Mystery', 'Thriller', 'Science Fiction']  
...                                                 ...  
9990                               ['Romance', 'Drama']  
9991                              ['Comedy', 'Mystery']  
9992  ['Drama', 'Horror', 'Science Fiction', 'Thrill...  
9993                               ['Drama', 'Romance']  
9994                                         ['Horror']  

[9995 rows x 2 columns]

In [42]:
df.columns

Index(['id', 'title', 'release_date', 'genres', 'original_language',
       'vote_average', 'vote_count', 'popularity', 'budget',
       'production_companies', 'revenue', 'runtime', 'year'],
      dtype='object')

## Data Cleaning Methods
## Data cleaning means fixing or removing incorrect, missing, or unwanted data in your DataFrame. It helps make your data accurate and ready for analysis.

In [ ]:
'''
21.dropna()
df.dropna()
Drops rows with missing values.
'''
df.dropna()

In [ ]:
'''
22.dropna() - for columns
df.dropna(axis=1)
Drops columns with missing values.
'''
df.dropna(axis = 1)

In [ ]:
'''
23.fillna()
df.fillna(0)
Fills the missing values with 0
'''
df.fillna(0)

In [ ]:
'''
24.fillna() - forward fill
df.fillna(method='ffill', inplace=True)
Fills missing values with the value from the previous row.
'''
df.fillna(method = 'ffill',inplace = True)

In [ ]:
'''
25.fillna() - Backward fill
df.fillna(method='bfill', inplace=True)
Fills missing values with the value from the next row.
'''
df.fillna(method = 'bfill', inplace = True)

In [ ]:
'''
26.fillna() - using mean
df['column_name'] = df['column_name'].fillna(df['column_name'].mean())
Fills missing values with the column mean
'''
df['mood_score'] = df['mood_score'].fillna(df['mood_score'].mean())

In [ ]:
'''
27.fillna() - using median
df['column_name'] = df['column_name'].fillna(df['column_name'].median())
Fills missing values with the column median
'''
df['meditation_sessions'] = df['meditation_sessions'].fillna(df['meditation_sessions'].median())

In [ ]:
'''
28.fillna() - using mode
df['column_name'] = df['column_name'].fillna(df['column_name'].mode()[0])
Fills missing values with the column median
mode() returns a Series. Use [0] to get the most frequent value.
'''
df['meditation_sessions'] = df['meditation_sessions'].fillna(df['meditation_sessions'].mode()[0])

In [ ]:
'''
29.fillna - different values/methods for different columns
df.fillna({
    'column1': value_or_method1,
    'column2': value_or_method2,
    ...
}, inplace=True)
fills missing values with different values for different columns.
'''
df.fillna({
    'original_language' : 

In [45]:
'''
30.replace()
df['column'].replace({old_value: new_value, ...}, inplace=True)
'''

df['original_language'].replace({3: 10}, inplace=True)



In [46]:
'''
31.rename() 
df.rename(columns={'old_name': 'new_name'}, inplace=True)
Renames the column name with the specified new name.
'''
df.rename(columns={'original_language': 'language'}, inplace=True) 


In [59]:
'''
32.reset_index()
df.reset_index(level=None, drop=False, inplace=False)
level: Resets a specific level if MultiIndex. (Optional)
drop: True = remove index (don’t add it as a column)
False = add index as a column                      
inplace: True = modify the original DataFrame
False = return a new DataFrame
It is used to reset the index of a DataFrame back to the default integer index
'''
df.reset_index(level=None, drop=False, inplace=False)


index       id                                      title release_date  \
0         4   536437                                   Hypnotic   2023-05-11   
1         6   890771                            The Black Demon   2023-04-26   
2        12  1010581                                   My Fault   2023-06-08   
3        15   879444             Accident Man: Hitman's Holiday   2022-10-14   
4        17  1073140                       The Three Musketeers   2023-03-23   
...     ...      ...                                        ...          ...   
4388   9978   575091                                   Tantrica   2018-11-30   
4389   9980    32093                          The Mark of Zorro   1940-11-08   
4390   9981    18222              Poison Ivy: The New Seduction   1997-10-20   
4391   9987   396398  The Meyerowitz Stories (New and Selected)   2017-10-13   
4392   9989    12720                               Suicide Club   2001-10-29   

                                           genres  language  vote_average  \
0      ['Mystery', 'Thriller', 'Science Fiction']   English           6.5   
1                          ['Horror', 'Thriller']   English           6.5   
2                            ['Romance', 'Drama']   Spanish           8.3   
3                   ['Action', 'Comedy', 'Crime']   English           6.6   
4                         ['Action', 'Adventure']   English           5.4   
...                                           ...       ...           ...   
4388                      ['Fantasy', 'Thriller']   English           5.0   
4389  ['Action', 'Adventure', 'Drama', 'Romance']   English           7.1   
4390             ['Thriller', 'Romance', 'Drama']   English           4.8   
4391                          ['Comedy', 'Drama']   English           6.5   
4392              ['Drama', 'Horror', 'Thriller']  Japanese           6.6   

      vote_count  popularity      budget  \
0          154.0    2654.854  70000000.0   
1          151.0    1777.200         0.0   
2          313.0    1170.670         0.0   
3           79.0    1117.559         0.0   
4           25.0     988.223         0.0   
...          ...         ...         ...   
4388         5.0      12.309    360000.0   
4389       174.0      12.309         0.0   
4390        82.0      12.309         0.0   
4391      1066.0      12.303         0.0   
4392       411.0      12.302    250000.0   

                                   production_companies  revenue  runtime  \
0     ['Studio 8', 'Solstice Productions', 'Ingeniou...      0.0     94.0   
1     ['Mucho Mas Media', 'Silk Mass', 'The Avenue E...      0.0    100.0   
2                 ['Pokeepsie Films', 'Amazon Studios']      0.0    117.0   
3                                   ['Six Demon Films']      0.0     96.0   
4                                                    []      0.0     94.0   
...                                                 ...      ...      ...   
4388                    ['Nighthawk Productions Assen']      0.0     44.0   
4389                               ['20th Century Fox']      0.0     94.0   
4390  ['New Line Cinema', 'CineTel Pictures', 'MG En...      0.0     93.0   
4391                   ['Gilded Halfwing', 'IAC Films']      0.0    112.0   
4392                                  ['Omega Project']      0.0     99.0   

        year  
0     2023.0  
1     2023.0  
2     2023.0  
3     2022.0  
4     2023.0  
...      ...  
4388  2018.0  
4389  1940.0  
4390  1997.0  
4391  2017.0  
4392  2001.0  

[4393 rows x 14 columns]

In [62]:
'''
33.str.strip()
df['column'] = df['column'].str.strip()
Removes Leading/Trailing Spaces in Strings.
'''
df['language'] = df['language'].astype(str).str.strip()


In [63]:
'''
34.notnull()
df = df[df['column'].notnull()]
Drops rows with incorrect data
'''
df = df.dropna(subset=['language'])


In [65]:
'''
35.str.lower()/str.upper()
df['column'] = df['column'].str.lower()
df['column'] = df['column'].str.upper()
Converts text in a column to lowercase or uppercase
'''
df['language'] = df['language'].str.upper()


## Handling Outliers in Pandas.

In [52]:
'''
36.Simple filtering using threshold
df = df[df['column'] < upper_limit]
df = df[df['column'] > lower_limit]
Remove values beyond a fixed limit.
'''
df = df[df['revenue'] < 8]
print(df) 

           id                                      title release_date  \
4      536437                                   Hypnotic   2023-05-11   
6      890771                            The Black Demon   2023-04-26   
12    1010581                                   My Fault   2023-06-08   
15     879444             Accident Man: Hitman's Holiday   2022-10-14   
17    1073140                       The Three Musketeers   2023-03-23   
...       ...                                        ...          ...   
9978   575091                                   Tantrica   2018-11-30   
9980    32093                          The Mark of Zorro   1940-11-08   
9981    18222              Poison Ivy: The New Seduction   1997-10-20   
9987   396398  The Meyerowitz Stories (New and Selected)   2017-10-13   
9989    12720                               Suicide Club   2001-10-29   

                                           genres  language  vote_average  \
4      ['Mystery', 'Thriller', 'Science Fictio

## Data Transformation Methods .

In [53]:
'''
37.map()
df['column'].map(function_or_dict)
It is used to map or transform values in a single column or Series.
'''
df['revenue'].map(lambda x : x + 5) 

4       5.0
6       5.0
12      5.0
15      5.0
17      5.0
       ... 
9978    5.0
9980    5.0
9981    5.0
9987    5.0
9989    5.0
Name: revenue, Length: 4393, dtype: float64

In [80]:
df.columns

Index(['id', 'title', 'release_date', 'genres', 'original_language',
       'vote_average', 'vote_count', 'popularity', 'budget',
       'production_companies', 'revenue', 'runtime', 'year'],
      dtype='object')

In [56]:
'''
38.sort_values()
df.sort_values(by='column_name', ascending=True)
by: Column name(s) to sort by
ascending: True = ascending order (default), False = descending
inplace: True to modify the original DataFrame
It sorts the DataFrame by column values
'''
df.sort_values(by = 'budget', ascending = False)

id                                          title release_date  \
294   335977          Indiana Jones and the Dial of Destiny   2023-06-28   
568   575264  Mission: Impossible - Dead Reckoning Part One   2023-07-10   
6407   83533                                       Avatar 3   2024-12-18   
5910  216527                                       Avatar 4   2026-12-16   
7156  393209                                       Avatar 5   2028-12-20   
...      ...                                            ...          ...   
9905  361931                                        Ronaldo   2015-11-09   
9904  767377                                        Dhamaka   2021-11-19   
1228  883502                         Fortress: Sniper's Eye   2022-04-29   
1227  628878                                         Panama   2022-03-17   
9896  765869                                   Black Friday   2021-11-19   

                                                 genres language  \
294                             ['Adventure', 'Action']  English   
568                 ['Action', 'Adventure', 'Thriller']  English   
6407         ['Action', 'Science Fiction', 'Adventure']  English   
5910  ['Action', 'Science Fiction', 'Adventure', 'Fa...  English   
7156         ['Action', 'Adventure', 'Science Fiction']  English   
...                                                 ...      ...   
9905                         ['Documentary', 'History']  English   
9904                     ['Thriller', 'Drama', 'Crime']    Hindi   
1228                    ['Action', 'Thriller', 'Crime']  English   
1227                             ['Thriller', 'Action']  English   
9896            ['Horror', 'Comedy', 'Science Fiction']  English   

      vote_average  vote_count  popularity       budget  \
294            0.0         0.0     107.803  300000000.0   
568            0.0         0.0      69.750  290000000.0   
6407           0.0         0.0      16.071  250000000.0   
5910           0.0         0.0      16.850  250000000.0   
7156           0.0         0.0      15.027  250000000.0   
...            ...         ...         ...          ...   
9905           6.9       322.0      12.364          0.0   
9904           7.0        74.0      12.366          0.0   
1228           5.5       134.0      42.989          0.0   
1227           5.2       101.0      42.995          0.0   
9896           5.0       182.0      12.372          0.0   

                                   production_companies  revenue  runtime  \
294                                  ['Lucasfilm Ltd.']      0.0    154.0   
568         ['Paramount', 'Skydance', 'TC Productions']      0.0    163.0   
6407  ['Lightstorm Entertainment', '20th Century Stu...      0.0      0.0   
5910  ['Lightstorm Entertainment', '20th Century Stu...      0.0      0.0   
7156  ['Lightstorm Entertainment', '20th Century Stu...      0.0      0.0   
...                                                 ...      ...      ...   
9905  ['On the Corner Films', 'We Came, We Saw, We C...      0.0     92.0   
9904  ['RSVP Movies', 'Ram Madhvani Films', 'Globalg...      0.0    104.0   
1228    ['EFO Films', 'Grindstone Entertainment Group']      0.0     87.0   
1227  ['Yale Productions', 'SSS Entertainment', 'LB ...      0.0    100.0   
9896    ['The Warner Davis Company', 'MFW Productions']      0.0     84.0   

        year  
294   2023.0  
568   2023.0  
6407  2024.0  
5910  2026.0  
7156  2028.0  
...      ...  
9905  2015.0  
9904  2021.0  
1228  2022.0  
1227  2022.0  
9896  2021.0  

[4393 rows x 13 columns]

In [57]:
'''
39.sort_index()
df.sort_index(axis=0, ascending=True)
axis=0 → sort by row index
axis=1 → sort by column names
ascending: default is True
Sorts the DataFrame by row or column index
'''
df.sort_index(axis = 0, ascending = False)

id                                      title release_date  \
9989    12720                               Suicide Club   2001-10-29   
9987   396398  The Meyerowitz Stories (New and Selected)   2017-10-13   
9981    18222              Poison Ivy: The New Seduction   1997-10-20   
9980    32093                          The Mark of Zorro   1940-11-08   
9978   575091                                   Tantrica   2018-11-30   
...       ...                                        ...          ...   
17    1073140                       The Three Musketeers   2023-03-23   
15     879444             Accident Man: Hitman's Holiday   2022-10-14   
12    1010581                                   My Fault   2023-06-08   
6      890771                            The Black Demon   2023-04-26   
4      536437                                   Hypnotic   2023-05-11   

                                           genres  language  vote_average  \
9989              ['Drama', 'Horror', 'Thriller']  Japanese           6.6   
9987                          ['Comedy', 'Drama']   English           6.5   
9981             ['Thriller', 'Romance', 'Drama']   English           4.8   
9980  ['Action', 'Adventure', 'Drama', 'Romance']   English           7.1   
9978                      ['Fantasy', 'Thriller']   English           5.0   
...                                           ...       ...           ...   
17                        ['Action', 'Adventure']   English           5.4   
15                  ['Action', 'Comedy', 'Crime']   English           6.6   
12                           ['Romance', 'Drama']   Spanish           8.3   
6                          ['Horror', 'Thriller']   English           6.5   
4      ['Mystery', 'Thriller', 'Science Fiction']   English           6.5   

      vote_count  popularity      budget  \
9989       411.0      12.302    250000.0   
9987      1066.0      12.303         0.0   
9981        82.0      12.309         0.0   
9980       174.0      12.309         0.0   
9978         5.0      12.309    360000.0   
...          ...         ...         ...   
17          25.0     988.223         0.0   
15          79.0    1117.559         0.0   
12         313.0    1170.670         0.0   
6          151.0    1777.200         0.0   
4          154.0    2654.854  70000000.0   

                                   production_companies  revenue  runtime  \
9989                                  ['Omega Project']      0.0     99.0   
9987                   ['Gilded Halfwing', 'IAC Films']      0.0    112.0   
9981  ['New Line Cinema', 'CineTel Pictures', 'MG En...      0.0     93.0   
9980                               ['20th Century Fox']      0.0     94.0   
9978                    ['Nighthawk Productions Assen']      0.0     44.0   
...                                                 ...      ...      ...   
17                                                   []      0.0     94.0   
15                                  ['Six Demon Films']      0.0     96.0   
12                ['Pokeepsie Films', 'Amazon Studios']      0.0    117.0   
6     ['Mucho Mas Media', 'Silk Mass', 'The Avenue E...      0.0    100.0   
4     ['Studio 8', 'Solstice Productions', 'Ingeniou...      0.0     94.0   

        year  
9989  2001.0  
9987  2017.0  
9981  1997.0  
9980  1940.0  
9978  2018.0  
...      ...  
17    2023.0  
15    2022.0  
12    2023.0  
6     2023.0  
4     2023.0  

[4393 rows x 13 columns]

In [58]:
'''
40.groupby()
df.groupby('col').sum()
df.groupby(['col1', 'col2'])['val'].agg(['mean', 'max'])
The groupby() method takes one or more column names as arguments, which are used to group the DataFrame.
After grouping, you can apply various functions to each group.
Common aggregation functions include sum(), mean(), median(), count(), min(), max(), and std().
'''
df.groupby(['revenue','budget'])['vote_count'].agg(['mean','max'])

mean     max
revenue budget                          
0.0     0.0           261.986814  7958.0
        1.0           126.333333   207.0
        4.0          1043.000000  1043.0
        10.0            0.000000     0.0
        20.0          113.000000   113.0
...                          ...     ...
        250000000.0     0.000000     0.0
        290000000.0     0.000000     0.0
        300000000.0     0.000000     0.0
1.0     5.0           238.000000   238.0
7.0     0.0             5.000000     5.0

[263 rows x 2 columns]

In [66]:
df.isnull().sum()

id                       0
title                    0
release_date            19
genres                   0
language                 0
vote_average             0
vote_count               0
popularity               0
budget                   0
production_companies     0
revenue                  0
runtime                  0
year                    19
dtype: int64

In [68]:
df.describe(include= 'all')

id  title                   release_date     genres  \
count   4.393000e+03   4393                           4374       4393   
unique           NaN   4333                            NaN       1279   
top              NaN  Porno                            NaN  ['Drama']   
freq             NaN      4                            NaN        303   
mean    4.798175e+05    NaN  2011-03-12 23:59:20.493827072        NaN   
min     1.080000e+02    NaN            1902-04-17 00:00:00        NaN   
25%     6.585100e+04    NaN            2007-01-28 00:00:00        NaN   
50%     4.872420e+05    NaN            2018-05-18 12:00:00        NaN   
75%     8.000890e+05    NaN            2022-02-26 12:00:00        NaN   
max     1.136631e+06    NaN            2028-12-20 00:00:00        NaN   
std     3.688667e+05    NaN                            NaN        NaN   

       language  vote_average   vote_count   popularity        budget  \
count      4393   4393.000000  4393.000000  4393.000000  4.393000e+03   
unique       47           NaN          NaN          NaN           NaN   
top     ENGLISH           NaN          NaN          NaN           NaN   
freq       2497           NaN          NaN          NaN           NaN   
mean        NaN      5.899954   315.388573    32.106831  2.755751e+06   
min         NaN      0.000000     0.000000    12.302000  0.000000e+00   
25%         NaN      5.500000    25.000000    14.211000  0.000000e+00   
50%         NaN      6.400000   136.000000    17.608000  0.000000e+00   
75%         NaN      7.000000   357.000000    26.626000  0.000000e+00   
max         NaN     10.000000  9008.000000  2654.854000  3.000000e+08   
std         NaN      1.920674   573.513709    77.482788  1.525999e+07   

       production_companies      revenue      runtime         year  
count                  4393  4393.000000  4393.000000  4374.000000  
unique                 3290          NaN          NaN          NaN  
top                      []          NaN          NaN          NaN  
freq                    509          NaN          NaN          NaN  
mean                    NaN     0.001821    89.512634  2010.689300  
min                     NaN     0.000000     0.000000  1902.000000  
25%                     NaN     0.000000    82.000000  2007.000000  
50%                     NaN     0.000000    93.000000  2018.000000  
75%                     NaN     0.000000   105.000000  2022.000000  
max                     NaN     7.000000   247.000000  2028.000000  
std                     NaN     0.106682    31.116429    16.568168

In [69]:
df.head()

id                           title release_date  \
4    536437                        Hypnotic   2023-05-11   
6    890771                 The Black Demon   2023-04-26   
12  1010581                        My Fault   2023-06-08   
15   879444  Accident Man: Hitman's Holiday   2022-10-14   
17  1073140            The Three Musketeers   2023-03-23   

                                        genres language  vote_average  \
4   ['Mystery', 'Thriller', 'Science Fiction']  ENGLISH           6.5   
6                       ['Horror', 'Thriller']  ENGLISH           6.5   
12                        ['Romance', 'Drama']  SPANISH           8.3   
15               ['Action', 'Comedy', 'Crime']  ENGLISH           6.6   
17                     ['Action', 'Adventure']  ENGLISH           5.4   

    vote_count  popularity      budget  \
4        154.0    2654.854  70000000.0   
6        151.0    1777.200         0.0   
12       313.0    1170.670         0.0   
15        79.0    1117.559         0.0   
17        25.0     988.223         0.0   

                                 production_companies  revenue  runtime  \
4   ['Studio 8', 'Solstice Productions', 'Ingeniou...      0.0     94.0   
6   ['Mucho Mas Media', 'Silk Mass', 'The Avenue E...      0.0    100.0   
12              ['Pokeepsie Films', 'Amazon Studios']      0.0    117.0   
15                                ['Six Demon Films']      0.0     96.0   
17                                                 []      0.0     94.0   

      year  
4   2023.0  
6   2023.0  
12  2023.0  
15  2022.0  
17  2023.0

In [70]:
df.index

Index([   4,    6,   12,   15,   17,   18,   19,   20,   23,   27,
       ...
       9969, 9970, 9971, 9973, 9976, 9978, 9980, 9981, 9987, 9989],
      dtype='int64', length=4393)